src/virtual_rules.ml

```diff
@@ -70,23 +70,36 @@ module Gen (P : sig val sctx : Super_context.t end) = struct
                 ~name:(Lib.name vlib) in
             (vlib_modules, virtual_modules)
         in
-        let (missing_modules, impl_modules_with_intf) =
-          Module.Name.Map.foldi virtual_modules ~init:([], [])
-            ~f:(fun m _ (mms, ims) ->
+        let (missing_modules, impl_modules_with_intf, private_virtual_modules) =
+          Module.Name.Map.foldi virtual_modules ~init:([], [], [])
+            ~f:(fun m _ (mms, ims, pvms) ->
               match Module.Name.Map.find modules m with
-              | None -> (m :: mms, ims)
+              | None -> (m :: mms, ims, pvms)
               | Some m ->
-                ( mms
-                , if Module.has_intf m then
+                let ims =
+                  if Module.has_intf m then
                     Module.name m :: ims
                   else
                     ims
-                ))
+                in
+                let pvms =
+                  if Module.is_public m then
+                    pvms
+                  else
+                    Module.name m :: pvms
+                in
+                (mms, ims, pvms))
         in
         let module_list ms =
           List.map ms ~f:Module.Name.to_string
           |> String.concat ~sep:"\n"
         in
+        if private_virtual_modules <> [] then begin
```

Comment:
Since you're testing all 3 lists for emptiness, what do you think about refactoring into this:

```
let partitioned_modules = List.fold_left ... in
match partitioned_modules with
| _::_, _::_, _::_ -> {Implementation. ... }
| [], _, _ -> (* error 1 *)
| _, [], _ -> (* error 2 *)
| _, _, [] -> (* error 3 *)
```

(bonus points if the pattern matching can be extracted into a `check` function a bit higher)

Reply:
I think extracting the check into its own function but I'm not convinced the pattern matching is adding here. Now if we want to add another field, we'll have to update 3 other completely independent clauses. Why do we prefer the pattern matching here? I don't see any gains from exhaustivity checks here for example.